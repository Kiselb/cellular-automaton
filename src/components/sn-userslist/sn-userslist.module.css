.panel {
    margin-bottom: 1.25em;
}
.user {
    display: flex;
}
.mark {
    display: flex;
    flex-direction: row;
    justify-content: space-around;
    width: 3.25em;
    height: 3.25em;
    border-style: none;
    border-radius: 50%;
    background: green;
    margin-right: 0.75em;
    color: white;
    font-size: 1em;
}
.mark > div {
    font-size: 2em;
    cursor: pointer;
}
.username {
    font-size: 1.5em;
    cursor: pointer;
}
