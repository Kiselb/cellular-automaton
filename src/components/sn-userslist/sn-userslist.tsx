import React, { Component } from "react";
import { TUserProps, User } from "../sn-user/sn-user";
import styles from "./sn-userslist.module.css";

type TUsersListProps = {
  userSelected: (userId: number) => any;
  probe?: (param: any) => any;
};
type TUsersList = {
  users: TUserProps[];
  displayUser: string[];
};

export class UsersList extends Component<TUsersListProps, TUsersList> {
  constructor(props: TUsersListProps) {
    super(props);
    this.state = {
      users: [],
      displayUser: [],
    };
    this.viewDetails = this.viewDetails.bind(this);
    this.onClickUserName = this.onClickUserName.bind(this);
  }
  shouldComponentUpdate(nextProps: TUsersListProps, nextState: TUsersList) {
    return (
      this.state.users.length !== nextState.users.length ||
      this.state.displayUser.length !== nextState.displayUser.length
    );
  }
  componentDidMount() {
    console.log(`Get Users`);
    fetch("https://jsonplaceholder.typicode.com/users")
      .then((response) => response.json())
      .then((result) => {
        this.setState({
          users: result.sort((a: TUserProps, b: TUserProps) => {
            if (a.username > b.username) {
              return 1;
            }
            if (a.username < b.username) {
              return -1;
            }
            return 0;
          }),
        });
      })
      .catch((error) => {
        console.log(error);
        this.setState({ users: [] });
      });
  }
  viewDetails(username: string) {
    !!this.props.probe && this.props.probe(null);
    return () =>
      this.setState((prevState) => {
        if (prevState.displayUser.indexOf(username) < 0) {
          return { displayUser: [...prevState.displayUser, username] };
        } else {
          return {
            displayUser: prevState.displayUser.filter(
              (item) => item !== username
            ),
          };
        }
      });
  }
  onClickUserName(userId: number) {
    return () => this.props.userSelected(userId);
  }
  render() {
    return (
      <>
        {this.state.users.map((user) => (
          <div key={user.id} className={styles.panel}>
            <div className={styles.user}>
              <div
                className={styles.mark}
                onClick={this.viewDetails(user.username)}
                data-testid="40"
              >
                <div>...</div>
              </div>
              <div>
                <div
                  className={styles.username}
                  onClick={this.onClickUserName(user.id)}
                  data-testid="50"
                >
                  {user.username}
                </div>
                <div>{user.name}</div>
              </div>
            </div>
            {this.state.displayUser.indexOf(user.username) >= 0 ? (
              <User {...user} />
            ) : null}
          </div>
        ))}
      </>
    );
  }
}
