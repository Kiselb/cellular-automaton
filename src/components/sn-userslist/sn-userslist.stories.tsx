import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";

import { UsersList } from "./sn-userslist";

export default {
  title: "Messenger/User",
  component: UsersList,
  argTypes: {
    userSelected: { action: "User Selected" },
  },
} as ComponentMeta<typeof UsersList>;

const Template: ComponentStory<typeof UsersList> = (args) => (
  <UsersList {...args} />
);

export const Users = Template.bind({});
Users.args = {};
