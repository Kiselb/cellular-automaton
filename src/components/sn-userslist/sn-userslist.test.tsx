import React from "react";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";

import "@testing-library/jest-dom";
import { UsersList } from "./sn-userslist";

it("Users List renders correctly", async () => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve([
          {
            id: 1,
            name: "Leanne Graham",
            username: "Bret",
            email: "Sincere@april.biz",
            address: {
              street: "Kulas Light",
              suite: "Apt. 556",
              city: "Gwenborough",
              zipcode: "92998-3874",
              geo: {
                lat: "-37.3159",
                lng: "81.1496",
              },
            },
            phone: "1-770-736-8031 x56442",
            website: "hildegard.org",
            company: {
              name: "Romaguera-Crona",
              catchPhrase: "Multi-layered client-server neural-net",
              bs: "harness real-time e-markets",
            },
          },
          {
            id: 2,
            name: "Ervin Howell",
            username: "Antonette",
            email: "Shanna@melissa.tv",
            address: {
              street: "Victor Plains",
              suite: "Suite 879",
              city: "Wisokyburgh",
              zipcode: "90566-7771",
              geo: {
                lat: "-43.9509",
                lng: "-34.4618",
              },
            },
            phone: "010-692-6593 x09125",
            website: "anastasia.net",
            company: {
              name: "Deckow-Crist",
              catchPhrase: "Proactive didactic contingency",
              bs: "synergize scalable supply-chains",
            },
          },
          {
            id: 3,
            name: "Leanne Graham",
            username: "Bret",
            email: "Sincere@april.biz",
            address: {
              street: "Kulas Light",
              suite: "Apt. 556",
              city: "Gwenborough",
              zipcode: "92998-3874",
              geo: {
                lat: "-37.3159",
                lng: "81.1496",
              },
            },
            phone: "1-770-736-8031 x56442",
            website: "hildegard.org",
            company: {
              name: "Romaguera-Crona",
              catchPhrase: "Multi-layered client-server neural-net",
              bs: "harness real-time e-markets",
            },
          },
          {
            id: 4,
            name: "Xeanne Graham",
            username: "Bret",
            email: "Sincere@april.biz",
            address: {
              street: "Kulas Light",
              suite: "Apt. 556",
              city: "Gwenborough",
              zipcode: "92998-3874",
              geo: {
                lat: "-37.3159",
                lng: "81.1496",
              },
            },
            phone: "1-770-736-8031 x56442",
            website: "hildegard.org",
            company: {
              name: "Romaguera-Crona",
              catchPhrase: "Multi-layered client-server neural-net",
              bs: "harness real-time e-markets",
            },
          },
        ]),
    })
  ) as jest.Mock;

  const { asFragment } = await waitFor(() =>
    render(<UsersList userSelected={() => console.log("OK")} />)
  );
  screen.debug();

  expect(asFragment()).toMatchSnapshot();
});
it("Users List Error catched", async () => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () => Promise.reject("Test Error"),
    })
  ) as jest.Mock;

  const { asFragment } = await waitFor(() =>
    render(<UsersList userSelected={() => console.log("OK")} />)
  );
  screen.debug();

  expect(asFragment()).toMatchSnapshot();
});
it("Users List Select User", async () => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve([
          {
            id: 1,
            name: "Leanne Graham",
            username: "Bret",
            email: "Sincere@april.biz",
            address: {
              street: "Kulas Light",
              suite: "Apt. 556",
              city: "Gwenborough",
              zipcode: "92998-3874",
              geo: {
                lat: "-37.3159",
                lng: "81.1496",
              },
            },
            phone: "1-770-736-8031 x56442",
            website: "hildegard.org",
            company: {
              name: "Romaguera-Crona",
              catchPhrase: "Multi-layered client-server neural-net",
              bs: "harness real-time e-markets",
            },
          },
        ]),
    })
  ) as jest.Mock;

  const handle = jest.fn();
  await waitFor(() => render(<UsersList userSelected={handle} />));

  const div = screen.getByTestId("50") as HTMLDivElement;
  fireEvent.click(div);
  expect(handle).toBeCalled();
});
it("Users List View Details request", async () => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve([
          {
            id: 1,
            name: "Leanne Graham",
            username: "Bret",
            email: "Sincere@april.biz",
            address: {
              street: "Kulas Light",
              suite: "Apt. 556",
              city: "Gwenborough",
              zipcode: "92998-3874",
              geo: {
                lat: "-37.3159",
                lng: "81.1496",
              },
            },
            phone: "1-770-736-8031 x56442",
            website: "hildegard.org",
            company: {
              name: "Romaguera-Crona",
              catchPhrase: "Multi-layered client-server neural-net",
              bs: "harness real-time e-markets",
            },
          },
        ]),
    })
  ) as jest.Mock;

  const handle1 = jest.fn();
  const handle2 = jest.fn();

  await waitFor(() =>
    render(<UsersList userSelected={handle1} probe={handle2} />)
  );

  const div1 = screen.getByTestId("40") as HTMLDivElement;
  fireEvent.click(div1);
  expect(handle2).toBeCalled();

  const div2 = screen.getByTestId("50") as HTMLDivElement;
  fireEvent.click(div2);
  expect(handle1).toBeCalled();

  fireEvent.click(div1);
  expect(handle2).toBeCalled();
});
