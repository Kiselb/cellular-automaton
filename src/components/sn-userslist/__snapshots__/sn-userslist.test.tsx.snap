// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users List Error catched 1`] = `<DocumentFragment />`;

exports[`Users List renders correctly 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div
        data-testid="40"
      >
        <div>
          ...
        </div>
      </div>
      <div>
        <div
          data-testid="50"
        >
          Antonette
        </div>
        <div>
          Ervin Howell
        </div>
      </div>
    </div>
  </div>
  <div>
    <div>
      <div
        data-testid="40"
      >
        <div>
          ...
        </div>
      </div>
      <div>
        <div
          data-testid="50"
        >
          Bret
        </div>
        <div>
          Leanne Graham
        </div>
      </div>
    </div>
  </div>
  <div>
    <div>
      <div
        data-testid="40"
      >
        <div>
          ...
        </div>
      </div>
      <div>
        <div
          data-testid="50"
        >
          Bret
        </div>
        <div>
          Leanne Graham
        </div>
      </div>
    </div>
  </div>
  <div>
    <div>
      <div
        data-testid="40"
      >
        <div>
          ...
        </div>
      </div>
      <div>
        <div
          data-testid="50"
        >
          Bret
        </div>
        <div>
          Xeanne Graham
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
